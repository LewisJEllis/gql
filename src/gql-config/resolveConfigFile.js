/* @flow */
import {
  type GQLConfigFile,
  type GQLConfigFileResolved,
  type ValidateConfig,
  type ValidateConfigResolved,
} from './types';
import { loadQueryPresets, loadSchemaPresets } from './loadPresets';
import { loadSchemaParser, loadQueryParser } from './loadParser';
import { mergeQueryPresets, mergeSchemaPresets } from './mergePresets';

// will resolve all packages in config and normalize the config file
export default function resolveConfigFile(
  config: GQLConfigFile,
  configFilePath: string,
): GQLConfigFileResolved {
  const schemaPresets = loadSchemaPresets(
    config.schema.presets,
    configFilePath,
  );
  const schemaConfig = mergeSchemaPresets(schemaPresets);

  const [SchemaParserClass, schemaParserOptions] = loadSchemaParser(
    schemaConfig.parser,
    configFilePath,
  );

  const schemaConfigResolved = {
    files: config.schema.files,
    validate: resolveValidateConfig(schemaConfig.validate),
    parser: [
      SchemaParserClass,
      {
        // NOTE: preset should come first
        ...schemaConfig.parserOptions,
        ...schemaParserOptions,
      },
    ],
  };

  if (!config.query) {
    return {
      schema: schemaConfigResolved,
    };
  }

  return {
    schema: schemaConfigResolved,
    query: {
      files: config.query.files.map(({ match, parser, presets, validate }) => {
        // load presets packages
        const queryPresets = loadQueryPresets(presets, configFilePath);
        // merge presets to generate config
        const presetConfig = mergeQueryPresets([
          ...queryPresets,
          { parser, validate },
        ]);

        const [ParserClass, parserOptions] = loadQueryParser(
          // NOTE: settings in .gqlconfig overrides preset
          presetConfig.parser,
          configFilePath,
        );

        return {
          match,
          validate: resolveValidateConfig(presetConfig.validate),
          extendSchema: presetConfig.extendSchema,
          parser: [
            ParserClass,
            {
              ...presetConfig.parserOptions,
              ...parserOptions,
            },
          ],
        };
      }),
    },
  };
}

function resolveValidateConfig(
  validateConfig: ?ValidateConfig,
): ValidateConfigResolved {
  if (!validateConfig) {
    return {
      rules: [],
      config: {},
    };
  }

  const { rules, config } = validateConfig;
  return {
    rules: rules
      ? Object.keys(rules).reduce((acc, ruleName) => {
          const rule = rules[ruleName];
          if (config[ruleName] && config[ruleName] !== 'off') {
            acc.push(rule);
          }
          return acc;
        }, [])
      : [],
    config: validateConfig.config,
  };
}
