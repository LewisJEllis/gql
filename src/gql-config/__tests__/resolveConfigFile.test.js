/* @flow */
import resolveConfigFile from '../resolveConfigFile';
import { createTempFiles } from 'gql-test-utils/file';

describe('schema-config', () => {
  test('default case', () => {
    expect(
      resolveConfigFile(
        {
          schema: {
            files: 'file/schema/*.gql',
          },
        },
        '',
      ),
    ).toMatchSnapshot();
  });

  test('use custom validation config', () => {
    expect(
      resolveConfigFile(
        {
          schema: {
            files: 'file/schema/*.gql',
            validate: {
              rules: {
                NoUnusedTypeDefinition: 'error',
              },
            },
          },
        },
        '',
      ),
    ).toMatchSnapshot();
  });

  test('use custom parser', () => {
    const rootPath = createTempFiles({
      './custom-parser.js': `
        module.exports = function customParser() {};
      `,
    });

    const resolvedConfig = resolveConfigFile(
      {
        schema: {
          files: 'file/schema/*.gql',
          parser: ['./custom-parser', { customParserOption: 5 }],
        },
      },
      rootPath,
    );
    expect(resolvedConfig).toMatchSnapshot();
    expect(resolvedConfig.schema.parser[0].name).toEqual('customParser');
  });

  test('use custom preset', () => {
    const rootPath = createTempFiles({
      'custom-preset': `
        module.exports = function custonPreset() {
          return {
            parser: 'gql-schema-parser-custom',
            parserOptions: { 'some-option': true },
            validate: {
              rules: [
                function someCustomRule() {},
              ],
              config: {
                someCustomRule: 'warn',
              },
            }
          };
        };
      `,
      'node_modules/gql-schema-parser-custom/package.json': `
        {
          "name": "gql-schema-parser-custom",
          "version": "1.0",
          "main": "index.js"
        }
      `,
      'node_modules/gql-schema-parser-custom/index.js': `
        module.exports = function customParser() {};
      `,
    });

    const resolvedConfig = resolveConfigFile(
      {
        schema: {
          files: 'file/schema/*.gql',
          preset: './custom-preset',
        },
      },
      rootPath,
    );

    expect(resolvedConfig).toMatchSnapshot();
  });
});

describe('query-files', () => {
  test('default case', () => {
    expect(
      resolveConfigFile(
        {
          schema: {
            files: 'file/schema/*.gql',
          },
          query: {
            files: [
              {
                match: 'files/**.js',
              },
            ],
          },
        },
        '',
      ),
    ).toMatchSnapshot();
  });

  describe('preset', () => {
    ['relay', 'apollo'].forEach(preset => {
      test(`preset: ${preset}`, () => {
        expect(
          resolveConfigFile(
            {
              schema: {
                files: 'file/schema/*.gql',
              },
              query: {
                files: [
                  {
                    match: 'files/**.js',
                    preset,
                  },
                ],
              },
            },
            '',
          ),
        ).toMatchSnapshot();
      });
    });

    test('custom-preset', () => {
      const rootPath = createTempFiles({
        'custom-preset': `
          module.exports = function customPreset() {
            return {
              parser: 'gql-query-parser-custom',
              parserOptions: { 'some-option': true },
              validate: {
                rules: [
                  function someCustomRule() {},
                ],
                config: {
                  someCustomRule: 'warn',
                },
              }
            };
          };
        `,
        'node_modules/gql-query-custom/package.json': `
          {
            "name": "gql-query-parser-custom",
            "version": "1.0",
            "main": "index.js"
          }
        `,
        'node_modules/gql-query-parser-custom/index.js': `
          module.exports = function customParser() {};
        `,
      });

      const resolvedConfig = resolveConfigFile(
        {
          schema: {
            files: 'file/schema/*.gql',
          },
          query: {
            files: [
              {
                match: 'files/**.js',
                preset: './custom-preset',
              },
            ],
          },
        },
        rootPath,
      );

      expect(resolvedConfig).toMatchSnapshot();
    });
  });
});
