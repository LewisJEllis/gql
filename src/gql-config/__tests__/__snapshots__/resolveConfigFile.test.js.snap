// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query-files default case 1`] = `
Object {
  "query": Object {
    "files": Array [
      Object {
        "extendSchema": undefined,
        "match": "files/**.js",
        "parser": Array [
          [Function],
          Object {},
        ],
        "validate": Object {
          "config": Object {
            "ArgumentsOfCorrectType": "error",
            "DefaultValuesOfCorrectType": "error",
            "FieldsOnCorrectType": "error",
            "FragmentsOnCompositeTypes": "error",
            "KnownArgumentNames": "error",
            "KnownDirectives": "error",
            "KnownFragmentNames": "error",
            "KnownTypeNames": "error",
            "LoneAnonymousOperation": "error",
            "NoFragmentCycles": "error",
            "NoUndefinedVariables": "error",
            "NoUnusedFragments": "error",
            "NoUnusedVariables": "error",
            "OverlappingFieldsCanBeMerged": "error",
            "PossibleFragmentSpreads": "error",
            "ProvidedNonNullArguments": "error",
            "RequiredOperationName": "warn",
            "ScalarLeafs": "error",
            "UniqueArgumentNames": "error",
            "UniqueDirectivesPerLocation": "error",
            "UniqueFragmentNames": "error",
            "UniqueInputFieldNames": "error",
            "UniqueOperationNames": "error",
            "UniqueVariableNames": "error",
            "VariablesAreInputTypes": "error",
            "VariablesInAllowedPosition": "error",
          },
          "rules": Array [
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
          ],
        },
      },
    ],
  },
  "schema": Object {
    "files": "file/schema/*.gql",
    "parser": Array [
      [Function],
      Object {},
    ],
    "validate": Object {
      "config": Object {
        "NoUnusedTypeDefinition": "warn",
      },
      "rules": Array [
        [Function],
      ],
    },
  },
}
`;

exports[`query-files preset custom-preset 1`] = `
Object {
  "query": Object {
    "files": Array [
      Object {
        "extendSchema": undefined,
        "match": "files/**.js",
        "parser": Array [
          [Function],
          Object {
            "some-option": true,
          },
        ],
        "validate": Object {
          "config": Object {
            "someCustomRule": "warn",
          },
          "rules": Array [
            [Function],
          ],
        },
      },
    ],
  },
  "schema": Object {
    "files": "file/schema/*.gql",
    "parser": Array [
      [Function],
      Object {},
    ],
    "validate": Object {
      "config": Object {
        "NoUnusedTypeDefinition": "warn",
      },
      "rules": Array [
        [Function],
      ],
    },
  },
}
`;

exports[`query-files preset preset: apollo 1`] = `
Object {
  "query": Object {
    "files": Array [
      Object {
        "extendSchema": undefined,
        "match": "files/**.js",
        "parser": Array [
          [Function],
          Object {
            "allowDocumentInterpolation": true,
            "end": "\`",
            "start": "gql\`",
          },
        ],
        "validate": Object {
          "config": Object {
            "ArgumentsOfCorrectType": "error",
            "DefaultValuesOfCorrectType": "error",
            "FieldsOnCorrectType": "error",
            "FragmentsOnCompositeTypes": "error",
            "KnownArgumentNames": "error",
            "KnownDirectives": "error",
            "KnownFragmentNames": "off",
            "KnownTypeNames": "error",
            "LoneAnonymousOperation": "off",
            "NoFragmentCycles": "error",
            "NoUndefinedVariables": "error",
            "NoUnusedFragments": "off",
            "NoUnusedVariables": "error",
            "OverlappingFieldsCanBeMerged": "error",
            "PossibleFragmentSpreads": "error",
            "ProvidedNonNullArguments": "error",
            "RequiredOperationName": "warn",
            "ScalarLeafs": "error",
            "UniqueArgumentNames": "error",
            "UniqueDirectivesPerLocation": "error",
            "UniqueFragmentNames": "error",
            "UniqueInputFieldNames": "error",
            "UniqueOperationNames": "error",
            "UniqueVariableNames": "error",
            "VariablesAreInputTypes": "error",
            "VariablesInAllowedPosition": "error",
          },
          "rules": Array [
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
          ],
        },
      },
    ],
  },
  "schema": Object {
    "files": "file/schema/*.gql",
    "parser": Array [
      [Function],
      Object {},
    ],
    "validate": Object {
      "config": Object {
        "NoUnusedTypeDefinition": "warn",
      },
      "rules": Array [
        [Function],
      ],
    },
  },
}
`;

exports[`query-files preset preset: relay 1`] = `
Object {
  "query": Object {
    "files": Array [
      Object {
        "extendSchema": [Function],
        "match": "files/**.js",
        "parser": Array [
          [Function],
          Object {
            "allowFragmentInterpolation": true,
            "allowFragmentWithoutName": true,
            "end": "\`",
            "start": "Relay\\\\.QL\`",
          },
        ],
        "validate": Object {
          "config": Object {
            "ArgumentsOfCorrectType": "error",
            "DefaultValuesOfCorrectType": "off",
            "FieldsOnCorrectType": "error",
            "FragmentsOnCompositeTypes": "error",
            "KnownArgumentNames": "error",
            "KnownDirectives": "error",
            "KnownFragmentNames": "off",
            "KnownTypeNames": "error",
            "LoneAnonymousOperation": "off",
            "NoFragmentCycles": "off",
            "NoUndefinedVariables": "off",
            "NoUnusedFragments": "off",
            "NoUnusedVariables": "off",
            "OverlappingFieldsCanBeMerged": "error",
            "PossibleFragmentSpreads": "error",
            "ProvidedNonNullArguments": "error",
            "RequiredOperationName": "warn",
            "ScalarLeafs": "error",
            "UniqueArgumentNames": "error",
            "UniqueDirectivesPerLocation": "error",
            "UniqueFragmentNames": "off",
            "UniqueInputFieldNames": "error",
            "UniqueOperationNames": "off",
            "UniqueVariableNames": "error",
            "UniqueVariablesName": "off",
            "VariablesAreInputTypes": "off",
            "VariablesInAllowedPosition": "off",
          },
          "rules": Array [
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
            [Function],
          ],
        },
      },
    ],
  },
  "schema": Object {
    "files": "file/schema/*.gql",
    "parser": Array [
      [Function],
      Object {},
    ],
    "validate": Object {
      "config": Object {
        "NoUnusedTypeDefinition": "warn",
      },
      "rules": Array [
        [Function],
      ],
    },
  },
}
`;

exports[`schema-config default case 1`] = `
Object {
  "schema": Object {
    "files": "file/schema/*.gql",
    "parser": Array [
      [Function],
      Object {},
    ],
    "validate": Object {
      "config": Object {
        "NoUnusedTypeDefinition": "warn",
      },
      "rules": Array [
        [Function],
      ],
    },
  },
}
`;

exports[`schema-config use custom preset 1`] = `
Object {
  "schema": Object {
    "files": "file/schema/*.gql",
    "parser": Array [
      [Function],
      Object {
        "some-option": true,
      },
    ],
    "validate": Object {
      "config": Object {
        "someCustomRule": "warn",
      },
      "rules": Array [
        [Function],
      ],
    },
  },
}
`;

exports[`schema-config use custom parser 1`] = `
Object {
  "schema": Object {
    "files": "file/schema/*.gql",
    "parser": Array [
      [Function],
      Object {
        "customParserOption": 5,
      },
    ],
    "validate": Object {
      "config": Object {
        "NoUnusedTypeDefinition": "warn",
      },
      "rules": Array [
        [Function],
      ],
    },
  },
}
`;

exports[`schema-config use custom validation config 1`] = `
Object {
  "schema": Object {
    "files": "file/schema/*.gql",
    "parser": Array [
      [Function],
      Object {},
    ],
    "validate": Object {
      "config": Object {
        "NoUnusedTypeDefinition": "error",
      },
      "rules": Array [
        [Function],
      ],
    },
  },
}
`;
